### PR TITLE
fwd freerdp logs to logging

### PR DESCRIPTION
**Pull request by mihainradu**

### CODE REVIEW

File: UiPath.FreeRdpClient/UiPath.FreeRdpClient/Internals/Logging.cs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        NativeInterface.InitializeLogging(logCallback: LogCallbackDelegate,
+                                          registerThreadScopeCallback: RegisterThreadScopeCallbackDelegate,
+                                          forwardFreeRdpLogs: forwardFreeRdpLogs);
+    }
+
+    private static void RegisterThreadScope(string scope)
+    {
+        BeginScope(scope);
+
+        static void BeginScope(string scopeValue)
```
Comment (eduard-dumitru) [inline UiPath.FreeRdpClient/UiPath.FreeRdpClient/Internals/Logging.cs:79]: I don't get it. Why do we need this local function?

File: UiPath.FreeRdpClient/UiPath.FreeRdpWrapper/FreeRdpWrapper.cpp [hunk 2]
```diff
@@ -16,6 +16,7 @@ namespace FreeRdpClient
 	{
 		rdpContext* context;
 		HANDLE transportStopEvent;
+		LPWSTR loggingScope;
```
Comment (eduard-dumitru) [inline UiPath.FreeRdpClient/UiPath.FreeRdpWrapper/FreeRdpWrapper.cpp:19]: Where is this used?

File: UiPath.FreeRdpClient/UiPath.FreeRdpClient/FreeRdpClient.cs [hunk 3]
```diff
@@ -69,49 +30,47 @@ public static async Task<IAsyncDisposable> Connect(RdpConnectionSettings connect
             Port = connectionSettings.Port ?? default
         };
 
-        return await Task.Run(() =>
+        return await Task.Run(async () =>
```
Comment (eduard-dumitru) [inline UiPath.FreeRdpClient/UiPath.FreeRdpClient/FreeRdpClient.cs:33]: Why did this lambda become async?

File: UiPath.FreeRdpClient/UiPath.FreeRdpClient/FreeRdpClient.cs [hunk 4]
```diff
@@ -69,49 +30,47 @@ public static async Task<IAsyncDisposable> Connect(RdpConnectionSettings connect
...
         {
-            RdpLogon(connectOptions, out var releaseObjectName);
-            return new AsyncDisposable(() => { Disconnect(releaseObjectName); return ValueTask.CompletedTask; });
+            NativeInterface.RdpLogon(connectOptions, out var releaseObjectName);
+            return new AsyncDisposable(() =>
+            {
+                Disconnect(releaseObjectName); 
+                return ValueTask.CompletedTask;
```
Comment (eduard-dumitru) [inline UiPath.FreeRdpClient/UiPath.FreeRdpClient/FreeRdpClient.cs:39]: I have two concerns here:
- Dispose and DisposeAsync should never throw.
- If Task-like returning method throws, it should return a faulted task instead of failing to return one.

I would suggest we do at least the following:

```csharp
return new AsyncDisposable(() =>
{
    try 
    {
        Disconnect(releaseObjectName); 
        return ValueTask.CompletedTask;
    }
    catch (Exception ex)
    {
       return ValueTask.FromException(ex);
    }
});
```

### Explanation

In the callee, we do:

```cpp
...
if (!SetEvent(eventHandle))
{
    auto lastError = GetLastError();
    CloseHandle(eventHandle);
    return HRESULT_FROM_WIN32(lastError);
}
...
```

which through PreserveSig=false would lead to `Disconnect(releaseObjectName)` throwing.

File: UiPath.FreeRdpClient/UiPath.FreeRdpClient/FreeRdpClient.cs [hunk 5]
```diff
@@ -69,49 +30,47 @@ public static async Task<IAsyncDisposable> Connect(RdpConnectionSettings connect
...
-
-    public static void SetupLogging(ILogger? logger)
-    {
-        Logger = logger;
-        InitializeLogging(LogCallbackDelegate);
-    }
-
     private static void Disconnect(string releaseObjectName)
     {
         if (releaseObjectName != default)
```
Comment (eduard-dumitru) [inline UiPath.FreeRdpClient/UiPath.FreeRdpClient/FreeRdpClient.cs:46]: **nit**

```csharp
if (releaseObjectName is not null)
```

File: UiPath.FreeRdpClient/UiPath.FreeRdpClient/Internals/Logging.cs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        if (!FilterLogs(logLevel, message))
+            return;
+
+        var log = LoggerFactory.CreateLogger(category);
+        log.Log(logLevel, message);
+    }
+
+    public static void SetupLogging(ILoggerFactory? loggerFactory)
+    {
+        var forwardFreeRdpLogs = Environment.GetEnvironmentVariable("WLOG_FILEAPPENDER_OUTPUT_FILE_PATH") is null;
```
Comment (eduard-dumitru) [inline UiPath.FreeRdpClient/UiPath.FreeRdpClient/Internals/Logging.cs:68]: `"WLOG_FILEAPPENDER_OUTPUT_FILE_PATH"` could be promoted to a public const in a Constants class.

File: UiPath.FreeRdpClient/UiPath.FreeRdpClient/UiPath.FreeRdpClient.csproj [hunk 7]
```diff
@@ -12,6 +12,7 @@
 	<ItemGroup>
 		<PackageReference Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="6.0.0" />
 		<PackageReference Include="Microsoft.Extensions.Hosting.Abstractions" Version="6.0.0" />
+		<PackageReference Include="Microsoft.Extensions.Logging" Version="6.0.0" />
```
Comment (eduard-dumitru) [inline UiPath.FreeRdpClient/UiPath.FreeRdpClient/UiPath.FreeRdpClient.csproj:15]: Why is this needed?

File: UiPath.FreeRdpClient/UiPath.FreeRdpClient/UiPath.FreeRdpClient.csproj [hunk 8]
```diff
@@ -12,6 +12,7 @@
 	<ItemGroup>
 		<PackageReference Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="6.0.0" />
```
Comment (eduard-dumitru) [inline UiPath.FreeRdpClient/UiPath.FreeRdpClient/UiPath.FreeRdpClient.csproj:13]: This is included with the next PackageReference: Microsoft.Extensions.Hosting.Abstractions

**Review by eduard-dumitru — CHANGES_REQUESTED** [review]
For now, I left out the filtering part.

File: UiPath.FreeRdpClient/UiPath.FreeRdpClient/Internals/Logging.cs [hunk 9]
```diff
@@ -1,17 +1,18 @@
...
 namespace UiPath.Rdp;
 
-internal static class Logging 
+internal class Logging : IHostedService
 {
     public static string ScopeName = "RunId";
-    internal static NativeInterface.LogCallback LogCallbackDelegate = Log;
-    private static NativeInterface.RegisterThreadScopeCallback RegisterThreadScopeCallbackDelegate = RegisterThreadScope;
+    private static NativeInterface.LogCallback LogCallbackDelegate = null!;
+    private static NativeInterface.RegisterThreadScopeCallback RegisterThreadScopeCallbackDelegate = null!;
```
Comment (eduard-dumitru) [inline UiPath.FreeRdpClient/UiPath.FreeRdpClient/Internals/Logging.cs:11]: If the purpose was initially to avoid garbage collection, 
then these should become instance fields and get assigned in the instance constructor:

```csharp
    private readonly NativeInterface.LogCallback LogCallbackDelegate;
    private readonly NativeInterface.RegisterThreadScopeCallback RegisterThreadScopeCallbackDelegate;

   ...
   
    public Logging(ILoggerFactory loggerFactory)
    {
        LogCallbackDelegate = Log;
        RegisterThreadScopeCallbackDelegate = RegisterThreadScope;
    
        LoggerFactory = loggerFactory;        
    }
```

File: UiPath.FreeRdpClient/UiPath.FreeRdpClient/Internals/FreeRdpClient.cs [hunk 10]
```diff
@@ -1,20 +1,47 @@
...
     }
 
-    public static async Task<IAsyncDisposable> Connect(RdpConnectionSettings connectionSettings)
+    public FreeRdpClient(ILogger<FreeRdpClient> logger)
+    {
+        _log = logger;
+    }
+
+    private readonly ILogger<FreeRdpClient> _log;
+    private static AsyncLock? InitLock = new();
```
Comment (eduard-dumitru) [inline UiPath.FreeRdpClient/UiPath.FreeRdpClient/Internals/FreeRdpClient.cs:22]: 1/🧵: Why is this mutable?

File: UiPath.FreeRdpClient/UiPath.FreeRdpClient/Internals/FreeRdpClient.cs [hunk 11]
```diff
@@ -1,20 +1,47 @@
...
+            return null;
+
+        var releaseLock = await InitLock.LockAsync();
+        if (InitLock is null)
+            return null;
+
+        _log.LogInformation("RdpInitLock acquired.");
+        return Disposable.Create(() =>
+        {
+            InitLock = null;
```
Comment (eduard-dumitru) [inline UiPath.FreeRdpClient/UiPath.FreeRdpClient/Internals/FreeRdpClient.cs:38]: 3/🧵 Why do we forget this instance?

File: UiPath.FreeRdpClient/UiPath.FreeRdpClient/Internals/FreeRdpClient.cs [hunk 12]
```diff
@@ -1,20 +1,47 @@
...
+
+    private async Task<IDisposable?> TryAcquireInitLock()
+    {
+        /// Make sure freerdp static initilizers are not run concurrently
+        /// when they do they fail with 
+        if (InitLock is null)
+            return null;
+
+        var releaseLock = await InitLock.LockAsync();
+        if (InitLock is null)
```
Comment (eduard-dumitru) [inline UiPath.FreeRdpClient/UiPath.FreeRdpClient/Internals/FreeRdpClient.cs:32]: 2/🧵 I don't see the reason behind this mechanism, and I think we can fix whatever the problem differently.

⚠️ Should we not find a better solution:
Even though references are the same size as IntPtr which means that assigning is atomic, this doesn't fix the multicore with distinct caches problem, and we should still use synchronization.

**Review by eduard-dumitru — CHANGES_REQUESTED** [review]

File: UiPath.FreeRdpClient/UiPath.FreeRdpClient.Tests/RdpClientTests.cs [hunk 13]
```diff
@@ -71,6 +80,54 @@ public async Task ShouldConnect(int colorDepthInput, int expectedWtsApiValue)
...
+        };
+        var iterations = 10;
+
+        while (iterations-- > 0)
+        {
+            var freerdpAppDataFolder = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "freerdp");
+            if (Directory.Exists(freerdpAppDataFolder))
+                Directory.Delete(freerdpAppDataFolder, recursive: true);
+
+            freerdpAppDataFolder = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.SystemX86), "config\\systemprofile\\AppData\\Roaming\\freerdp");
```
Comment (vuplea) [inline UiPath.FreeRdpClient/UiPath.FreeRdpClient.Tests/RdpClientTests.cs:111]: These assignments are left-overs, or something, remove.

File: UiPath.FreeRdpClient/UiPath.FreeRdpClient.Tests/RdpClientTests.cs [hunk 14]
```diff
@@ -104,27 +108,26 @@ public async Task ParallelConnectWorksOnFirstUse()
...
                 Directory.Delete(freerdpAppDataFolder, recursive: true);
 
-            var connect1Task = Connect(connectionSettings1);
-            var connect2Task = Connect(connectionSettings2);
+            var connect1Task = host.GetFreeRdpClient().Connect(connectionSettings1);
+            var connect2Task = host.GetFreeRdpClient().Connect(connectionSettings2);
 
             await using var d = new CollectionAsyncDisposable(await Task.WhenAll(connect1Task, connect2Task));
 
             await d.DisposeAsync();
```
Comment (eduard-dumitru) [inline UiPath.FreeRdpClient/UiPath.FreeRdpClient.Tests/RdpClientTests.cs:128]: **super nit** double dispose 1 statement away

File: UiPath.FreeRdpClient/UiPath.FreeRdpClient/Internals/FreeRdpClient.cs [hunk 15]
```diff
@@ -56,14 +37,29 @@ public async Task<IAsyncDisposable> Connect(RdpConnectionSettings connectionSett
             HostName = connectionSettings.HostName,
             Port = connectionSettings.Port ?? default
         };
-        using var releaseInitLock = await TryAcquireInitLock();
-        return await Task.Run(async () =>
+
+        using (await _initLock.LockAsync())
+        {
+            /// Make sure freerdp static initilizers are not run concurrently
```
Comment (eduard-dumitru) [inline UiPath.FreeRdpClient/UiPath.FreeRdpClient/Internals/FreeRdpClient.cs:43]: **nit** typo (initi**a**lizers)

**Review by eduard-dumitru — APPROVED** [review]
